test/test_euler.py

```diff
@@ -0,0 +1,472 @@
+    rhoE = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    rhoV = make_obj_array(
+        [
+            cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+            for i in range(dim)
+        ]
+    )
+    scal1 = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    p = cl.clrandom.rand(queue, (mesh.nelements,), dtype=np.float64)
+    ke = 0.5 * np.dot(rhoV, rhoV) / rho
+
+    # ideal single spec
+    gamma = 1.4
+    p = (gamma - 1.0) * (rhoE - ke)
+    escale = (rhoE + p) / rho
+    expected_mass_flux = rhoV
+    expected_energy_flux = scalevec(escale, rhoV)
+    expected_mom_flux = make_obj_array(
+        [
+            (rhoV[i] * rhoV[j] / rho + (p if i == j else 0))
+            for i in range(dim)
+            for j in range(dim)
+        ]
+    )
+
+    q = join_fields(rho, rhoE, rhoV)
+
+    from mirgecom.euler import _inviscid_flux
+
+    flux = _inviscid_flux(discr, q)
+
+    rhoflux = flux[0:dim]
+    rhoEflux = flux[dim : 2 * dim]
+    momflux = flux[2 * dim :]
+
+    # these should be exact, right?
+    for i in range(dim):
+        assert (
+            la.norm(rhoflux[i].get() - expected_mass_flux[i].get())
+            == 0.0
+        )
+        assert (
+            la.norm(rhoEflux[i].get() - expected_energy_flux[i].get())
+            == 0.0
+        )
+    for i in range(dim * dim):
+        assert (
+            la.norm(momflux[i].get() - expected_mom_flux[i].get())
+            == 0.0
+        )
+
+
+def test_facial_flux():
+    cl_ctx = cl.create_some_context()
+    queue = cl.CommandQueue(cl_ctx)
+
+    dim = 2
+    nel_1d = 16
+
+    from meshmode.mesh.generation import generate_regular_rect_mesh
+
+    mesh = generate_regular_rect_mesh(
+        a=(-0.5,) * dim, b=(0.5,) * dim, n=(nel_1d,) * dim
+    )
+
+    order = 3
+
+    if dim == 2:
+        # no deep meaning here, just a fudge factor
+        dt = 0.75 / (nel_1d * order ** 2)
+    elif dim == 3:
+        # no deep meaning here, just a fudge factor
+        dt = 0.45 / (nel_1d * order ** 2)
+    else:
+        raise ValueError("don't have a stable time step guesstimate")
+
+    print("%d elements" % mesh.nelements)
+
+    discr = EagerDGDiscretization(cl_ctx, mesh, order=order)
+
+    mass_input = discr.zeros(queue)
+    energy_input = discr.zeros(queue)
+    mom_input = join_fields(
+        [discr.zeros(queue) for i in range(discr.dim)]
+    )
+
+    # This sets p = 1
+    mass_input[:] = 1.0
+    energy_input[:] = 2.5
+
+    fields = join_fields(mass_input, energy_input, mom_input)
+
+    from mirgecom.euler import _facial_flux
+    from mirgecom.euler import _interior_trace_pair
+
+    interior_face_flux = _facial_flux(
+        discr, w_tpair=_interior_trace_pair(discr, fields)
+    )
+
+    err = np.max(
+        np.array(
+            [
+                la.norm(interior_face_flux[i].get(), np.inf)
+                for i in range(0, 2)
+            ]
+        )
+    )
+    assert err < 1e-15
```

Comment:
FP tol

Reply:
I think I've fixed up all of these that make sense.